tr: Stream output instead of buffering

BenWiederhake opened the pull request:
This should lower memory consumption, and fixes OOM in some scenarios.

Before, buggy behavior:

```console
$ yes | tr -ds a b | head -n5  # GNU
y
y
y
y
y
$ yes | cargo run -- -ds a b | head -n5  # WARNING: THIS CAUSES OOM!
^C
[$? = 130]
$ yes yes | tr -s ye s | head -n5  # GNU
s
s
s
s
s
$ yes yes | cargo run -- -s ye s | head -n5  # WARNING: THIS CAUSES OOM!
^C
[$? = 130]
```

After:

```console
$ yes | cargo run -- -ds a b | head -n5
y
y
y
y
y
$ yes yes | cargo run -- -s ye s | head -n5
s
s
s
s
s
$
```

It might be tempting to factor out the common line `translate_input(&mut locked_stdin, &mut buffered_stdout, op);` from all if-branches. However, note that `op` has distinct types in each case, and virtualizing it (e.g. `let op: Box<dyn SymbolTranslator>`) would incur additional runtime costs. That cost seems unnecessary just to gain a little bit of beauty.

I'm not really sure how to test this. Ideally we could pipe in some data, and *NOT* close the stdin pipe, and see whether the program produces any partial output. However, I didn't see any methods/functions for doing so, and it sounds like it's too hard to get right to be worth it.

Maybe I could write a `BENCHMARKING.md` file, like for `shuf`? What do you suggest?

BenWiederhake commented:
Changes since last push:
- Nothing, I just want to re-run CI
- CI flaked in `test_split::test_round_robin_limited_file_descriptors`

<details><summary>Test output of flaky test `test_split::test_round_robin_limited_file_descriptors`</summary>

```
---- test_split::test_round_robin_limited_file_descriptors stdout ----
run: /target/i686-unknown-linux-gnu/debug/coreutils split -n r/40 onehundredlines.txt
thread 'test_split::test_round_robin_limited_file_descriptors' panicked at 'Command was expected to succeed.
stdout = 
 stderr = split: unable to open 'xaz'; aborting
', tests/by-util/test_split.rs:1646:10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: tests::common::util::CmdResult::success
             at ./tests/common/util.rs:388:9
```

Sadly, this does not show a backtrace of the failing command. Also, I cannot reproduce this problem. On my machine the command only needs 5 open files at any point in time, so `9` seems like this already includes a lot of wiggle room, so raising this even further is unlikely to fix the issue.

</details>

sylvestre commented on src/uu/tr/src/tr.rs:
```diff
@@ -139,17 +133,10 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
-                let squeeze_op = SqueezeOperation::new(set2, false);
-                translate_input(&mut reader, &mut buffered_stdout, squeeze_op);
-            }
+            let translate_op = TranslateOperation::new(set1, set2.clone(), complement_flag)?;
```
As you are working on it and understand the code, it would be nice to add some comments explaining what is doing because it is a bit cryptic if you aren't familiar with this :)

BenWiederhake replied:
Apparently you're right in that I understand the code, because I don't see the crypticness :D

Can you clarify which aspect is cryptic?

- `translate_op` and `squeeze_op` are the two transformations that are selected, but the names already say that.
- They must be applied in this order, but `chain` already says that.
- `SymbolTranslator::chain` does pretty much the same thing as [`Iterator::chain`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.chain), would a link to it help?
- And finally we transform the entire input with the combined operation, but the code already says that.

BenWiederhake replied:
Okay! I'll mostly copy-paste the above description.

BenWiederhake commented:
Changes since last push:
- Rebased on current master
- Added documentation for `SymbolTranslator::chain`, and a short explanation that the `*_op` variables are the translation operations, as requested by @sylvestre

BenWiederhake commented:
Changes since last push:
- Apparently, links in the documentation need `<…>` around them.

tertsdiepraam commented on src/uu/tr/src/operation.rs:
```diff
@@ -339,6 +339,32 @@ impl Sequence {
     fn translate(&mut self, current: u8) -> Option<u8>;
+
+    /// Takes two SymbolTranslators and creates a new SymbolTranslator over both in sequence.
+    /// This behaves pretty much identical to `Iterator::chain`:
+    /// <https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.chain>
```
```suggestion
    ///
    /// This behaves pretty much identical to [`Iterator::chain`].
```

The newline helps rustdoc with formatting and it can figure out the link with just `[]`.

tertsdiepraam reviewed (commented):
I like it! Just one suggestion on the link in the comment.

BenWiederhake commented:
Changes since last push(es):
- Rebased on current `main`
- Make documentation prettier. Thanks @tertsdiepraam!
- Learn the hard way that the "commit suggestion" button doesn't create nice fixups.
